ast/walk.go

```diff
@@ -72,6 +72,112 @@ func Walk(expr Expression, visitor func(Expression) bool) bool {
+	case *Parent:
+		count--
+		if count <= 0 {
+			return count, fmt.Errorf("Parent usage is invalid")
+		}
+		return count, nil
+	case *Everything, *This, *Ellipsis, *StringLiteral,
+		*FloatLiteral, *BooleanLiteral, *IntegerLiteral, *NullLiteral, *Attribute,
+		*ArrayTraversal, *Param, *FunctionParam:
+		return count, nil
+	case *Constraint:
+		return WalkAndFindParentUsageInvalid(count, e.Expression)
+	case *Subscript:
+		return WalkAndFindParentUsageInvalid(count, e.Value)
+	case *Range:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.Start); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.End)
+	case *BinaryOperator:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.RHS)
+	case *DotOperator:
+		// Chain the count between LHS and RHS sides of the dot operator.
+		c, err := WalkAndFindParentUsageInvalid(count, e.LHS)
+		if err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(c, e.RHS)
+	case *Group:
+		return WalkAndFindParentUsageInvalid(count, e.Expression)
+	case *PipeOperator:
+		if _, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return count, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.RHS)
+	case *PrefixOperator:
+		return WalkAndFindParentUsageInvalid(count, e.RHS)
+	case *PostfixOperator:
+		return WalkAndFindParentUsageInvalid(count, e.LHS)
+	case *Array:
+	case *Object:
+		for _, expr := range e.Expressions {
+			if c, err := WalkAndFindParentUsageInvalid(count, expr); err != nil {
+				return c, err
+			}
+		}
+		return count, nil
+	case *FunctionCall:
+		for _, arg := range e.Arguments {
+			if c, err := WalkAndFindParentUsageInvalid(count, arg); err != nil {
+				return c, err
+			}
+		}
+		return count, nil
+	case *FunctionPipe:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.Func)
+	case *Filter:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
+
+		return WalkAndFindParentUsageInvalid(count+1, e.Constraint)
+	case *Slice:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.Range)
+	case *Element:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.Idx)
+	case *Projection:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count+1, e.Object)
+	case *Tuple:
+		for _, member := range e.Members {
+			if c, err := WalkAndFindParentUsageInvalid(count, member); err != nil {
+				return c, err
+			}
+		}
+		return count, nil
+	}
+
+	panic(fmt.Sprintf("Unexpected AST expression of type %T: %[1]v", expr))
```

Comment:
Nit: A bit more idiomatic to return an error object here (but I mean, it's a panic so all hope is lost anyway).

```suggestion
	panic(fmt.Errorf("nexpected AST expression of type %T: %[1]v", expr))
```